Adding a Swift-to-Dart guide (#4427)
dart-lang/site-www: src/_guides/language/coming-from/swift-to-dart.md

{{site.alert.note}}
  Flutter uses the Dart language,
  so if you are coding Flutter, you might find
  [Flutter for iOS developers][] to be useful.
{{site.alert.end}}

[Flutter for iOS developers]: {{site.flutter-docs}}/get-started/flutter-for/ios-devs

## Conventions and linting

Swift and Dart both have linting tools to enforce
standard conventions. However, while Swift has
`SwiftLint` as a standalone tool,
Dart has official layout conventions and includes a
linter to make compliance effortless.
To customize the lint rules for your project,
follow the [Customizing static analysis][] instructions.
(Note that the IDE plugins for Dart and Flutter also
provide this functionality.)

[Customizing static analysis]: /guides/language/analysis-options

{{site.alert.secondary}}
  **Pro tip:** Dart provides [`dart fix`][],
  which finds and fixes errors found by the analyzer.
{{site.alert.end}}

[`dart fix`]: /tools/dart-fix

Dart also provides a code formatter,
which can automatically format any Dart project when
running `dart format` from the command line or through
the IDE. At one time, Flutter provided a `flutter format`
command, but that is deprecated. Use the `dart format`

> command instead.

Author: i'm ok either way

{{site.alert.secondary}}
  **Pro tip:** Dart supports optional trailing
  commas for any comma-separated values,
  such as function parameters or list items.
  This forces the formatter to place each item
  onto its own line, which helps with readability,
  especially when you have a lot of nested code
  (as can happen in Flutter layout code).

  For more information on using commas to make
  your code more readable, check out

>   [Using trailing commas][] on docs.flutter.dev.

Author: Yeah, I don't want to send them to flutter.com. But I'm fine with it being a link.

libraries, such as:

* Basic value types like
  * Numbers (`num`, `int`, `double`)
  * Strings (`String`)
  * Booleans (`bool`)
  * The value null (`Null`)
* Collections
  * Lists/arrays (`List`)
  * Sets (`Set`)
  * Maps/dictionaries (`Map`)

For more information, check out [Built-in types][]
in the Dart language tour.

[Built-in types]: /guides/language/language-tour#built-in-types

### Numbers

Dart defines three numeric types for holding numbers:

`num`
: A generic 64-bit number type.

`int`
: A platform-dependent integer number.
  In native code, it's a 64-bit two's complement integer.
  On the web, it's a non-fractional 64-bit,
  floating-point number.

`double`
: A 64-bit floating point number.

Unlike Swift, there are no specific types for
unsigned integers.

All these types are also classes in the Dart API.
Both the `int` and `double` types share `num`
as their parent class:

<img
  src="/assets/img/number-class-hierarchy.svg"
  alt="Object is the parent of num, which is the parent of int and double">

As number values are technically class instances,
they have the convenience of exposing their own
utility functions. Because of this, a `double` can,

> for example, be rounded up as follows:

Reviewer: The following code snippets are not talking about rounding up a double, but rather converting an integer to a double. Would be nice to include the rounding example though :)

Author: I'm going to leave this to you. I finished this article as a favor to Mit. But I've told Tony that once, I land this PR, it's all his.  :D

with a question mark at the end of the type.

### Null-aware operators

Dart supports several operators to deal with nullability.
The null coalescing operator (`??`),
and optional chaining operator (`?.`)
are available in Dart and operate the same as in Swift:

{:.include-lang}
```dart
a = a ?? b;
```

{:.include-lang}
```swift
let str: String? = nil
let count = str?.count ?? 0
```

Additionally, Dart offers a
[null safe version of the cascade operator][] (`?..`),
which ignores any operations when
the target object is `null`.
Dart also offers the null assignment operator (`??=`),
which Swift doesn't.
It allows you to only assign a value to a variable
with a nullable type when its current value is `null`.
It's expressed as `a ??= b;` and is shorthand for
the following:

{:.include-lang}
```dart
a = a ?? b;

// equivalent to:
a ??= b; 
```

{:.include-lang}
```swift
a = a ?? b
```

#### ! Operator (also called "force unwrap")

In cases where it's safe to assume that a
nullable variable or expression is, in fact, non-null,
it's possible to tell the compiler to repress
any compile-time errors.
This is done using the suffix `!` operator,
by placing it as a suffix to the expression.

> (Don't confuse this with Dart's "not" operator,
> which uses the same symbol:

Reviewer: There is no closing parentheses here:

```suggestion
(Don't confuse this with Dart's "not" operator,
which uses the same symbol):
```

keyword to delay initialization of `final` variables.
While you don't have to immediately initialize
the `final` variable when marking it as `late`,
it can still be initialized only once.
A second assignment results in a runtime error.

{:.include-lang}
```dart
late final int a;
a = 1;
a = 2; // Throws a runtime exception because
       // "a" is already initialized.
```

## Functions

Swift uses the `main.swift` file as an app's
entry point. Dart uses the `main` function
as an app's entry point. Every program must
have a `main` function to be executable. For example:

{:.include-lang}
```dart
void main() {
  // main function is the entry point
  print("hello world");
}
```

{:.include-lang}
```swift
// main.swift file is the entry point
print("hello world")
```

Dart doesn't support `Tuples` (though there are
[several tuple packages][] available on pub.dev).
In the case that a function needs to return
multiple values, you can wrap them in a collection,
like a list, set, or map,
or you can write a wrapper class where an instance
can be returned that contains these values.
More on this can be found in the sections on
[collections](#collections) and [classes](#classes).

> [several tuple packages]: {{site.dart-site.pub}}/packages?q=tuples

Reviewer: ```suggestion
[several tuple packages]: {{site.pub}}/packages?q=tuples
```

### Exception and error handling

As in Swift, Dart's functions and methods support
handling both [exceptions][] and [errors][].
Dart _errors_ typically represent programmer mistakes
or system failures like stack overflow.
Dart errors aren't supposed to be caught.
On the other hand, Dart _exceptions_ represent
a recoverable failure, and are intended to be caught.
For example, at runtime the code might try to access
a streaming feed, but instead receives an exception which,
if not caught, results in termination of the app.
You can manage exceptions in Dart by wrapping the
function call in a `try-catch` block.

[errors]: {{site.dart-api}}/dart-core/Error-class.html
[exceptions]: {{site.dart-api}}/dart-core/Exception-class.html

{:.include-lang}
```dart
try {
  // Create audio player object

>   audioPlayer = AVAudioPlayer(soundURL);

Reviewer: Following Effective Dart conventions, this would be `soundUrl`:

```suggestion
  audioPlayer = AVAudioPlayer(soundUrl);
```

            
  // Play the sound
  audioPlayer?.play();
}
catch {
  // Couldn't create audio player object, log the exception
  print('Couldn't create the audio player for %s',
     soundFilename);
}
```

Similarly, Swift uses a `do-try-catch` block.
For example:

{:.include-lang}
```swift
do {
  // Create audio player object
  audioPlayer = try AVAudioPlayer(contentsOf: soundURL)
            
  // Play the sound
  audioPlayer?.play()
}
catch {
  // Couldn't create audio player object, log the error
  print("Couldn't create the audio player for file \(soundFilename)")
}
```

You can use the `try-catch` block in both
synchronous and asynchronous Dart code.
For more information,

> see the docs for the [Error][errors] and
> [Exception][exceptions] classes.

Reviewer: ```suggestion
see the docs for the [`Error`][errors] and
[`Exception`][exceptions] classes.
```

in its functions. However, unlike Swift,
these aren't the default in Dart.
The default parameter type in Dart is a
positional parameter.

{:.include-lang}
```dart
int multiply(int a, int b) {
  return a * b;
}
```

The equivalent in Swift prefaces a parameter
with an underscore to remove the need
for an argument label.

{:.include-lang}
```swift
func multiply(_ a: Int, _ b: Int) -> Int {
  return a * b
}
```

When creating named parameters in Dart,
they must be defined separately in a
block of curly braces,
after any positional parameters:

{:.include-lang}
```dart
int multiply(int a, int b, {int c = 1, int d = 1}) {
  return a * b * c * d;
}

// Calling a function with both required and named parameters
multiply(3, 5); // 15
multiply(3, 5, c: 2); // 30
multiply(3, 5, d: 3); // 45
multiply(3, 5, c: 2, d: 3); // 90
```

{:.include-lang}
```swift
// The Swift equivalent
func multiply(_ a: Int, _ b: Int, c: Int = 1, d: Int = 1) -> Int {
  return a * b * c * d
}
```

Named parameters must have a default value,
be marked as nullable (which defaults to a `null` value,
if not explicitly provided), or be marked as `required`
for the function to compile.

> Nullable types and null safety](#null-safety)

Reviewer: ```suggestion
Nullable types and [null safety](#null-safety)
```

subtle differences, shown in the following example:

{:.include-lang}
```dart
final List<String> list1 = <String>['one', 'two', 'three']; // Initialize list and specify full type
final list2 = <String>['one', 'two', 'three']; // Initialize list using shorthand type
final list3 = ['one', 'two', 'three']; // Dart can also infer the type
```

{:.include-lang}
```swift
var list1: Array<String> = ["one", "two", "three"] // Initialize array and specify the full type
var list2: [String] = ["one", "two", "three"] // Initialize array using shorthand type
var list3 = ["one", "two", "three"] // Swift can also infer the type
```

The following code samples give an overview of the
basic actions that you can perform on a Dart `List`.
The first example shows how to retrieve a value
from a list using the `index` operator:

{:.include-lang}
```dart
final fruits = ['apple', 'orange', 'pear'];
final fruit = fruits[1];
```

To add a value to the back of the list,
use the `add` method. To add another `List`
use the `addAll` method:

{:.include-lang}
```dart
final fruits = ['apple', 'orange', 'pear'];
fruits.add('peach');
fruits.addAll(['kiwi', 'mango']);
```

For the complete List API,

> refer to the [`List<E>` class] page.
> 
> [`List<E>` class]: {{site.dart-api}}/dart-core/List-class.html

Reviewer: I think it's fine to not include the generic here:

```suggestion
refer to the [`List` class] documentation.

[`List` class]: {{site.dart-api}}/dart-core/List-class.html
```

gifts.remove('first');
gifts.removeWhere((key, value) => value == 'partridge');
```

## Classes 

Dart doesn't define an interface type&mdash;_any_
class can be used as an interface.
If you want to introduce just an interface,
create an abstract class with no concrete members.
To get a more detailed understanding of these categories,
check out the documentation in the
[abstract classes](#abstract-classes),
[implicit interfaces](#implicit-interfaces),
and [extending a class](#extending-a-class) sections.

Dart doesn't provide support for value types.
As mentioned in the [Built-in types](#built-in-types) section,
all types in Dart are reference types (even primitives),
meaning that Dart doesn't provide a `struct` keyword.

### Enums

Enumerated types, often called enumerations or enums,
are a special kind of class used to represent
a fixed number of constant values. Enums have been
part of the Dart language for a long time,
but Dart 2.17 added enhanced enums support for members.
That means you can add fields holding state,
constructors that set that state,
methods with functionality,
and even override existing members.
For more information, check out

> [Enhanced enums with members][] on the Dart blog,
> and [Declaring enhanced enums][] in the Dart language tour.
> 
> [Declaring enhanced enums]: /guides/language/language-tour#declaring-enhanced-enums
> [Enhanced enums with members]: https://medium.com/dartlang/dart-2-17-b216bfc80c5d#:~:text=verbose%20and%20repetitive.-,Enhanced,-enums%20with%20members

Reviewer: I would rather just link to the language tour as the blog may get out of date and not all browsers support the text search syntax:

```suggestion
[Declaring enhanced enums][] in the Dart language tour.

[Declaring enhanced enums]: /guides/language/language-tour#declaring-enhanced-enums
```

}

// Create a new instance of the Point class
Point p = Point(3, 5);
```

Similar to functions, constructors can also take
optional positional or named parameters:

{:.include-lang}
```dart
class Point {
  ...
  // With an optional positioned parameter
  Point(this.x, [this.y = 0]);
  // With named parameters
  Point({required this.y, this.x = 0});
  // With both positional and named parameters
  Point(int x, int y, {int scale = 1}) {
    ...
  }
  ...
}
```

#### Initializer lists

You can also use initializer lists,
which run after any fields that are directly set
using `this` in the constructor parameters,
but still before the constructor body:

{:.include-lang}
```dart
class Point {
  ...
  Point(Map<String, double> json)
      : x = json['x']!,
        y = json['y']! {
    print('In Point.fromJson(): ($x, $y)');
  }
  ...
}
```

An initializer list is a good place to use an assert.

#### Named constructors

Unlike Swift, Dart allows classes to have multiple
constructors by allowing you to name them.
You have the option of using one unnamed constructor,
but any additional constructors must be named.
A class can also have only named constructors.

{:.include-lang}

> ```swift

Reviewer: Looks like this is meant to be Dart:

```suggestion
```dart
```

defines an interface containing all the
instance members of the class and of any
interface it implements. If you want to create
a class `A` that supports class `B`'s API
without inheriting `B`'s implementation,
class `A` should implement the `B` interface. 

Unlike Dart, Swift classes don't implicitly
define an interface. The interface needs to be explicitly
defined as a protocol and implemented by the developer.

A class can implement one or more interfaces
and then provide the APIs required by the interfaces.
Both Dart and Swift have a different ways of
implementing interfaces. For example:

{:.include-lang}
```dart
abstract class Animal {
  int getLegs();
  void makeNoise();
}

class Dog implements Animal {
  @override
  int getLegs() => 4;

  @override
  void makeNoise() => print('Woof woof');
}
```

{:.include-lang}
```swift
protocol Animal {
   func getLegs() -> Int;
   func makeNoise()
}

class Dog: Animal
{
  func getLegs() -> Int {
    return 4;
  }

  func makeNoise() {
    print("Woof woof"); 
  }
}
```

### Extending a class

Class inheritance in Dart is very similar to Swift.
In Dart, you can use `extends` to create a subclass,

> and `super` to refer to the superclass:

Reviewer: Note for others (or myself): Would be nice to include an example of using super in the following examples

var manager = NumberManager<String>(); 
```

### Generics in literals

`Map-`, `Set-`, and `List-` literals can explicitly
declare generic types, which is useful when the
type isn't inferred or is incorrectly inferred. 

For example, the `List` class has a generic definition:
`class List<E>`. Generic type `E` refers to the type
of the list's contents. Normally,
this type is automatically inferred,
which is used in some membery types of the `List` class's.
(For example, its first getter returns a value of type `E`).
When defining a `List` literal,
you can explicitly define the generic type as follows:

{:.include-lang}
```dart
var objList = [5, 2.0]; // Type: List<num> // Automatic type inference
var objList = <Object>[5, 2.0]; // Type: List<Object> // Explicit type definition
var objSet = <Object>{5, 2.0}; // Sets work identically
```

This is also true for a `Map`,
which also define its `key` and `value` types
using generics (`class Map<K, V>)`:

{:.include-lang}
```dart
// Automatic type inference
var map = {
  'foo': 'bar'
}; // Type: Map<String, String>
// Explicit type definition:
var map = <String, Object>{
  'foo': 'bar'
}; // Type: Map<String, Object>
```

## Concurrency

> Swift supports multithreading and Dart supports
> the `Isolate`, which are similar to lightweight threads

Reviewer: "the Isolate" sounds a bit weird to me:

```suggestion
Swift supports multithreading and Dart supports isolates, 
which are similar to lightweight threads
```

function and returns control to the event loop,
which can work on something else until the future
completes either with a value or an error.
At some time after that, the `await` expression
evaluates to that value or throws that error.

When it completes, the future's value is returned.
You can only `await` in an `async` context, as in Swift.

{:.include-lang}
```dart
// We can only await futures within an async context.
asyncFunction() async {
  String returnedString = await fetchString();
  print(returnedString); // 'String Value'
}
```

When the awaited future fails, an error object
is thrown on the line with the `await` keyword.
You can handle this using a regular `try-catch` block:

{:.include-lang}
```dart
// We can only await futures within an async context.
Future<void> asyncFunction() async {
  String? returnedString;
  try {
    returnedString = await fetchString();
  } catch (error) {
    print('Future encountered an error before resolving.');
    return;
  }
  print(returnedString);
}
```

For more information and some hands-on practice,
check out the [Asynchronous programming][] codelab.

[Asynchronous programming]: /codelabs/async-await

### Streams

> Another tool in Dart's async toolbox is the `Stream`.

Reviewer: ```suggestion
Another tool in Dart's async toolbox is the `Stream` class.
```

and returns a `Stream` instead of an `Iterable`.
This approach is similar to the `AsyncStream`
struct in Swift.

In an async generator function, the `yield` keyword
emits the given value to the stream.
The `yield*` keyword, however,
works with streams instead of other iterables.
This allows events from other streams to be
emitted to this stream. In the following example,
the function only continues once the newly yielded
stream has completed:

{:.include-lang}
```dart
Stream<int> asynchronousNaturalsTo(int n) async* {
  int k = 0;
  while (k < n) yield k++;
}

Stream<int> stream = asynchronousNaturalsTo(5); 
```

You can also create a stream using the
`StreamController` API. For more information,
see [Using a StreamController][].

[Using a StreamController]: /articles/libraries/creating-streams#using-a-streamcontroller

## Doc comments

Regular comments work the same in Dart as they do in Swift.
Using a double backslash (`//`) comments out everything
beyond the double slash for the rest of the line,
and `/* ... */` blocks comments spanning multiple lines.

In addition to regular comments,
Dart also has [doc comments][] that work in tandem

> with [dartdoc][]: a first-party tool that generates

Reviewer: ```suggestion
with [`dart doc`][]: a first-party tool that generates
```

are public by default. Private variables
are prefixed with the underline character (`_`),
and are enforced by the Dart compiler.

{:.include-lang}
```dart
final foo = 'this is a public property';
final _foo = 'this is a private property';

String bar() {
  return 'this is a public method';
}
String _bar() {
  return 'this is a private method';
}

// Public class
class Foo {
}

// Private class
class _Foo {
},
```

Private methods and variables are scoped
to their library in Dart, and to a module in Swift.
In Dart, you can define a library in a file,
while in Swift you must create a new build target
for your module. This means that in a single Dart
project you can define `n` libraries,
but in Swift you must create `n` modules. 

All files that are part of a library can gain access
to all private objects in that library.
But for security reasons, a file still needs
to allow specific files to gain access to its
private objects, otherwise any file&mdash;even from
outside of your project&mdash;could register itself to your
library and gain access to possibly sensitive data.
In other words,
private objects aren't shared across libraries.

**animal.dart**
{:.include-lang}
```dart

> library animals;

Reviewer: Consider adjusting this example to avoid specifying a named library and using a part file and instead using text to highlight they are in the same library of folder. Named libraries and parts are generally avoided now and not discussed elsewhere in this doc, so it may just cause unnecessary confusion.
